ops/model.py

```diff
@@ -1550,15 +1550,8 @@ def __iter__(self):
         return iter(self._data)
 
     def __getitem__(self, key: Union['Unit', 'Application']):
-        if key is None and self.relation.app is None:
-            # NOTE: if juju gets fixed to set JUJU_REMOTE_APP for relation-broken events, then that
-            # should fix the only case in which we expect key to be None - potentially removing the
-            # need for this error in future ops versions (i.e. if relation.app is guaranteed to not
-            # be None. See https://bugs.launchpad.net/juju/+bug/1960934.
-            raise KeyError(
-                'Cannot index relation data with "None".'
-                ' Are you trying to access remote app data during a relation-broken event?'
-                ' This is not allowed.')
+        if not isinstance(key, Unit) and not isinstance(key, Application):
```

Comment:
```suggestion
        if not isinstance(key, (Unit, Application)):
```

Reply:
I'm fine with this addition here for those not using static type checking, but I think in general our approach is to lean on static checking now.

Reply:
I felt this was better than checking explicitly for `None` as John suggested. I'm fine pulling it out completely if you prefer.

Reply:
Tony and I discussed, and we'd like our principle to be "we've invested in type annotations, let's go all in on static type checking and let that catch this." Most charming teams are using MyPy and PyRight already, and those that aren't, I think we should nudge them in that direction.